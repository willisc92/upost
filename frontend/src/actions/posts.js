import API, { setContentToForm, resetContentType } from "../utils/API";
import { getCurrentUser } from "./auth";
import moment from "moment";

/**
 * POST_START.
 * action generator.
 * Resets the error and sets loading status.
 *
 * @returns {Object} object to reset error and loading status
 */
export const postStart = () => ({
    type: "POST_START"
});

/**
 * SET_POSTS.
 * action generator.
 * adds user's posts to store.
 *
 * @param {Object[]} posts posts to add to store
 * @returns {Object} object to add posts to store
 */
export const setPosts = (posts) => ({
    type: "SET_POSTS",
    posts
});

/**
 * START_SET_POSTS.
 * from API gets the current user's posts and adds to store.
 *
 * @returns dispatch function
 */
export const startSetMyPosts = () => {
    return (dispatch) => {
        dispatch(postStart());
        getCurrentUser()
            .then((res) => {
                API.get("posts/", {
                    params: {
                        user: res.data.id
                    }
                })
                    .then((result) => {
                        dispatch(postSuccess());
                        dispatch(setPosts(result.data));
                    })
                    .catch((err) => {
                        dispatch(postFail(err));
                    });
            })
            .catch((err) => {
                console.log(JSON.stringify(err));
            });
    };
};

/**
 * SET_INTEREST_RANDOM_POSTS.
 * action generator.
 * resets the interestRandomPosts in store to be an empty array.
 *
 * @returns {Object} object to reset interestRandomPosts
 */
export const setInterestRandomPosts = () => {
    return { type: "SET_INTEREST_RANDOM_POSTS" };
};

/**
 * ADD_INTEREST_RANDOM_POSTS.
 * action generator.
 * adds a new interest and associated random posts to store.
 *
 * @param {Object} interestPosts contains tag and an array of posts
 * @returns {Object} object to add interest and posts to interestRandomPosts
 */
export const addInterestRandomPosts = (interestPosts) => {
    return { type: "ADD_INTEREST_RANDOM_POSTS", interestPosts };
};

/**
 * START_SET_INTEREST_RANDOM_POSTS.
 * resets the interestRandomPosts store.
 * for each interest provided gets random posts from API and adds to store.
 *
 * @param {Object[]} interests an array of interest objects or interest tags
 * @returns {Function} dispatch function
 */
export const startSetInterestRandomPosts = (interests) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            dispatch(setInterestRandomPosts());
            // if we have a list of interest objects
            if (interests.length > 0 && !!interests[0].interest_tag) {
                for (let i = 0; i < interests.length; i++) {
                    API.get("random-posts/", {
                        params: {
                            interest: interests[i].interest_tag
                        }
                    }).then((result) => {
                        if (result.data.length > 0) {
                            dispatch(addInterestRandomPosts({ tag: interests[i].interest_tag, posts: result.data }));
                        }
                    });
                }
            }
            // if we have a list of interest tags
            else {
                for (let i = 0; i < interests.length; i++) {
                    API.get("random-posts/", {
                        params: {
                            interest: interests[i]
                        }
                    }).then((result) => {
                        if (result.data.length > 0) {
                            dispatch(addInterestRandomPosts({ tag: interests[i], posts: result.data }));
                        }
                    });
                }
            }
            dispatch(postSuccess());
            resolve(true);
        });
    };
};

/**
 * START_GET_POST.
 * from API gets a specifc post and adds to store.
 *
 * @param {number} id post id to get
 * @returns {Promise}
 */
export const startGetPost = (id) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            API.get("posts/", {
                params: {
                    post_id: id
                }
            })
                .then((result) => {
                    dispatch(postSuccess());
                    dispatch(setPosts(result.data));
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * GET_POST.
 * from API gets a specific post, but does not add to store.  Does not dispatch any actions.
 *
 * @param {number} id post id to get
 */
export const getPost = (id) => {
    return new Promise((resolve, reject) => {
        API.get("posts/", {
            params: {
                post_id: id
            }
        })
            .then((result) => {
                resolve(result.data);
            })
            .catch((err) => {
                reject(err);
            });
    });
};

/**
 * ADD_POST.
 * sends post request to API to add a new user created post.
 * upon sucessful post reloads user posts.
 *
 * @param {Object} post post object to add to API
 * @returns {Promise}
 */
export const addPost = (post) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            if (post.toString() == "[object FormData]") {
                setContentToForm();
            }
            dispatch(postStart());
            API.post(`posts/`, post)
                .then((result) => {
                    if (post.toString() == "[object FormData]") {
                        resetContentType();
                    }
                    dispatch(postSuccess());
                    resolve(result);
                })
                .catch((err) => {
                    resetContentType();
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * EDIT_POST.
 * sends put request to API to edit a user's post.
 * upon sucessful post reloads user posts.
 *
 * @param {number} id post id to edit
 * @param {Object} updates updates to be made to the post
 */
export const editPost = (id, updates) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            if (updates.toString() == "[object FormData]") {
                setContentToForm();
            }
            dispatch(postStart());
            API.patch(`posts/${id}/`, updates)
                .then((result) => {
                    if (updates.toString() == "[object FormData]") {
                        resetContentType();
                    }
                    dispatch(postSuccess());
                    resolve(result);
                })
                .catch((err) => {
                    resetContentType();
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * POST_FAIL.
 * action generator.
 * adds error to the store.
 *
 * @param {Object} error the error
 * @returns {Object} object to set error
 */
export const postFail = (error) => ({
    type: "POST_FAIL",
    error
});

/**
 * POST_SUCCESS.
 * action generator.
 * resets loading and error status.
 *
 * @returns {Object} object to reset loading and error status
 */
export const postSuccess = () => ({
    type: "POST_SUCCESS"
});

/**
 * CLEAR_POSTS.
 * action generator.
 * clears posts in store
 *
 * @returns {Object} object to reset posts
 */
export const clearPosts = () => ({
    type: "CLEAR_POSTS"
});

/**
 * searchPosts
 * action dispatcher - searches for posts
 * @param {string} text - text to search posts by
 * @returns {Promise} to be handled
 */
export const searchPosts = (text) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            API.get(`posts/?search=${text}`)
                .then((result) => {
                    dispatch(postSuccess());
                    dispatch(setPosts(result.data));
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * deletePost
 * action dispatcher - soft deletes a post
 * @param {number} id - id of post to delete
 * @param {Object} post - Post object to delete
 * @returns {Promise} to be handled
 */
export const deletePost = (id) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            const updates = {
                deleted_flag: true,
                deletion_date: moment().toDate()
            };
            API.patch(`posts/${id}/`, updates)
                .then((result) => {
                    dispatch(postSuccess());
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * restorePost
 * action dispatcher - restores a post
 * @param {number} id - id of post to restore
 * @param {Object} post - Post object to restore
 * @returns {Promise} to be handled
 */
export const restorePost = (id) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            const updates = {
                deleted_flag: false,
                deletion_date: null
            };
            API.patch(`posts/${id}/`, updates)
                .then((result) => {
                    dispatch(postSuccess());
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * getNonInterestPosts
 * action dispatcher - gets posts that are NOT relavent to your interests
 * @returns {Promise} to be handled
 */
export const getNonInterestPosts = () => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            API.get(`non-interest-posts/`)
                .then((result) => {
                    dispatch(postSuccess());
                    dispatch(setNonInterestPosts(result.data));
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));

                    reject(err);
                });
        });
    };
};

/**
 * setNonInterestPosts
 * action generator - sets posts not relavent to interests
 * @returns {Object} Action object
 */
export const setNonInterestPosts = (nonInterestPosts) => ({
    type: "SET_NON_INTEREST_POSTS",
    nonInterestPosts
});

/**
 * getRandomNonInterestPost
 * action dispatcher - Gets a random post that is not relavent to interests
 * @returns {Promise} Promise to handle
 */
export const getRandomNonInterestPost = () => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            API.get(`random-post/`)
                .then((result) => {
                    dispatch(postSuccess());
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * getCommunityPosts
 * action dispatcher - Gets posts only according to interests
 * @returns {Promise} Promise to handle
 */
export const getCommunityPosts = () => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            API.get(`community-posts/`)
                .then((result) => {
                    dispatch(postSuccess());
                    dispatch(setPosts(result.data));
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};
