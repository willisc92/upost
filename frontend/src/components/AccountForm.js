import React from "react";
import { connect } from "react-redux";
import { SingleDatePicker } from "react-dates";
import Select from "react-select";
import moment from "moment";
import API from "../utils/API";

class AccountForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            calendarFocused: false,
            date: moment(),
            birthDate: moment(),
            isLoaded: false,
            interests: [],
            error: undefined,
            selectedInterests: []
        };
    }

    onDateChange = (birthDate) => {
        if (birthDate) {
            this.setState(() => ({ birthDate }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onInterestChange = (selectedOption) => {
        this.setState({ selectedInterests: selectedOption });
    };

    componentDidMount() {
        API.get("interests/").then(
            (result) => {
                const interests = result.data.map((interest) => {
                    return { value: interest.interest_tag, label: interest.interest_tag };
                });
                this.setState({
                    isLoaded: true,
                    interests: interests
                });
            },

            (error) => {
                this.setState({
                    isLoaded: true,
                    error
                });
            }
        );
    }

    onSubmit = (e) => {
        e.preventDefault();
        this.props.onSubmit({
            formInput: e.target,
            interests: this.state.selectedInterests.map((interest) => {
                return interest.value;
            }),
            date: this.state.date,
            birthDate: this.state.birthDate
        });
    };

    render() {
        return (
            <form onSubmit={this.onSubmit}>
                <p>Username*</p>
                {!!this.props.auth.error &&
                    !!this.props.auth.error.username &&
                    this.props.auth.error.username.map((error) => {
                        return <p key={error}>{error}</p>;
                    })}
                <input type="text" name="username" />
                <p>150 characters of fewer. Letters, digits and @/./+/-/) only.</p>
                <p>Email Address</p>
                {!!this.props.auth.error &&
                    !!this.props.auth.error.email &&
                    this.props.auth.error.email.map((error) => {
                        return <p key={error}>{error}</p>;
                    })}
                <input type="text" name="email" />
                <p>First Name*</p>
                {!!this.props.auth.error &&
                    !!this.props.auth.error.first_name &&
                    this.props.auth.error.first_name.map((error) => {
                        return <p key={error}>{error}</p>;
                    })}
                <input type="text" name="firstName" />
                <p>Middle Name</p>
                <input type="text" name="middleName" />
                <p>Last Name*</p>
                {!!this.props.auth.error &&
                    !!this.props.auth.error.last_name &&
                    this.props.auth.error.last_name.map((error) => {
                        return <p key={error}>{error}</p>;
                    })}
                <input type="text" name="lastName" />
                <p>Birth date</p>
                <SingleDatePicker
                    date={this.state.date}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={(day) => false}
                />
                <p>Country*</p>
                {!!this.props.auth.error &&
                    !!this.props.auth.error.country &&
                    this.props.auth.error.country.map((error) => {
                        return <p key={error}>{error}</p>;
                    })}
                <input type="text" name="country" />
                <p>State*</p>
                {!!this.props.auth.error &&
                    !!this.props.auth.error.state &&
                    this.props.auth.error.state.map((error) => {
                        return <p key={error}>{error}</p>;
                    })}
                <input type="text" name="state" />
                <p>Steet Name*</p>
                {!!this.props.auth.error &&
                    !!this.props.auth.error.street_name &&
                    this.props.auth.error.street_name.map((error) => {
                        return <p key={error}>{error}</p>;
                    })}
                <input type="text" name="streetName" />
                <p>Postal Code*</p>
                {!!this.props.auth.error &&
                    !!this.props.auth.error.postal_code &&
                    this.props.auth.error.postal_code.map((error) => {
                        return <p key={error}>{error}</p>;
                    })}
                <input type="text" name="postalCode" />
                <p>City*</p>
                {!!this.props.auth.error &&
                    !!this.props.auth.error.city &&
                    this.props.auth.error.city.map((error) => {
                        return <p key={error}>{error}</p>;
                    })}
                <input type="text" name="city" />
                <p>Sex</p>
                <select name="sex">
                    <option>Male</option>
                    <option>Female</option>
                </select>
                <p>Phone Number</p>
                <input type="text" name="phoneNumber" />
                <p>Interests*</p>
                {!!this.props.auth.error &&
                    !!this.props.auth.error.interests &&
                    this.props.auth.error.interests.map((error) => {
                        return <p key={error}>{error}</p>;
                    })}
                <Select
                    value={this.state.selectedInterests}
                    onChange={this.onInterestChange}
                    options={this.state.interests}
                    isMulti
                    name="interests"
                />
                <p>Password</p>
                <p>Your password contain at least {this.props.minPasswordLength} characters.</p>
                {!!this.props.auth.error &&
                    !!this.props.auth.error.password &&
                    this.props.auth.error.password.map((error) => {
                        return <p key={error}>{error}</p>;
                    })}
                <input type="text" name="password" />
                <p>Password Confirmation*</p>
                {!!this.props.auth.error &&
                    !!this.props.auth.error.password &&
                    this.props.auth.error.password.map((error) => {
                        return <p key={error}>{error}</p>;
                    })}
                <input type="text" name="passwordConfirmation" />
                <br />
                <button>Sign Up</button>
            </form>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        auth: state.auth
    };
};

export default connect(mapStateToProps)(AccountForm);
