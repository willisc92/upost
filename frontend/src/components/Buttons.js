import React from "react";
import Button from "@material-ui/core/Button";
import { makeStyles } from "@material-ui/core/styles";

export const useStyles = makeStyles((theme) => ({
    whiteButton: {
        color: "white"
    },
    whiteRedButton: {
        color: "white",
        backgroundColor: theme.palette.primary.main,
        margin: 5
    }
}));

const White = (WrappedComponent) => {
    const HOC = (props) => {
        const classes = useStyles();
        return <WrappedComponent {...props} className={classes.whiteButton} />;
    };

    return HOC;
};

const WhiteRed = (WrappedComponent) => {
    const HOC = (props) => {
        const classes = useStyles();
        return <WrappedComponent {...props} className={classes.whiteRedButton} />;
    };

    return HOC;
};

export const WhiteButton = White(Button);
export const WhiteRedButton = WhiteRed(Button);
