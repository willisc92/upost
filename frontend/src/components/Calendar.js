import { Calendar, momentLocalizer } from "react-big-calendar";
import moment from "moment";
import React from "react";
const localizer = momentLocalizer(moment);

export const MyCalendar = (props) => (
    <div className="calendar_container">
        <Calendar
            localizer={localizer}
            events={props.events}
            onSelectEvent={props.onSelectEvent}
            startAccessor="start"
            endAccessor="end"
            views={["month", "day", "week"]}
        />
    </div>
);
