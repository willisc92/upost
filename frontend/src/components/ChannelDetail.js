import React from "react";
import API from "../utils/API";

class ChannelDetail extends React.Component {
    state = {
        detail: {},
        isLoaded: false,
        error: null
    };

    componentDidMount() {
        API.get("channels/").then(
            (result) => {
                this.setState({
                    isLoaded: true,
                    channels: result.data
                });
            },

            (error) => {
                this.setState({
                    isLoaded: true,
                    error
                });
            }
        );
    }

    render() {
        const { error, isLoaded, items } = this.state;
        if (error) {
            return <div>Error: {error.message}</div>;
        } else if (!isLoaded) {
            return <div>Loading...</div>;
        } else {
            return (
                <div>
                    {this.state.channels.map((channel) => (
                        <div key={channel.channel_id}>
                            <h1>{channel.channel_name}</h1>
                            <span>{channel.creation_date}</span>
                        </div>
                    ))}
                </div>
            );
        }
    }
}

export default ChannelList;
