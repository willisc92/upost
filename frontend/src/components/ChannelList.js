import React from 'react';

class ChannelList extends React.Component {
    state = {
        channels: []
    };

    async componentDidMount() {
        try {
          const res = await fetch('http://127.0.0.1:8000/api/channels/');
          const channels = await res.json();
          this.setState({
            channels
          });
        } catch (e) {
          console.log(e);
        }
    }
    
    render() {
      return (
        <div>
          {this.state.channels.map((channel) => (
            <div key={channel.channel_id}>
              <h1>{channel.channel_name}</h1>
              <span>{channel.creation_date}</span>
            </div>
          ))}
        </div>
      );
    }
}

export default ChannelList; 

    