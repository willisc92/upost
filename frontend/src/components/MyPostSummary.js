import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";

export const MyPostSummary = ({ post, pathName, selected, readOnly }) => {
    return (
        <div className={`menu-item ${selected ? "active" : ""}`}>
            <Link
                className={readOnly ? "polaroid__inactive" : "polaroid"}
                to={{
                    pathname: pathName,
                    state: { post }
                }}
            >
                <div className="polaroid__text-wrapper">
                    <img className="polaroid__image" src={post.picture} />
                    <h3 className="polaroid__title">
                        {post.post_title}{" "}
                        {post.deleted_flag && (
                            <span className="polaroid__sub_title">
                                {" "}
                                (Deleted {moment(post.deletion_date).format("ddd, MMM D YYYY")})
                            </span>
                        )}
                    </h3>
                    <p className="polaroid__description">{post.post_description}</p>
                    <p className="polaroid__description">
                        Creation Date: {moment(post.post_timestamp).format("MMMM Do YYYY")}
                    </p>
                </div>
            </Link>
        </div>
    );
};

export default MyPostSummary;

export const MyPostMenu = (list, selected, readOnly) =>
    list.map((el) => {
        return (
            <MyPostSummary
                post={el}
                pathName={`/myPosts/${el.post_id}/edit`}
                key={el.post_id}
                selected={selected}
                readOnly={readOnly}
            />
        );
    });

export const BrowsePostMenu = (list, selected) =>
    list
        .filter((el) => !el.deleted_flag)
        .map((el) => {
            return <MyPostSummary post={el} pathName={`/post/${el.post_id}`} key={el.post_id} selected={selected} />;
        });
