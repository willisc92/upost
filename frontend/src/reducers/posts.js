// Post Reducers
export const postsDefaultState = { error: null, loading: false, posts: [], interestRandomPosts: [] };

export default (state = postsDefaultState, action) => {
    switch (action.type) {
        case "POST_START":
            return {
                ...state,
                error: null,
                loading: true
            };
        case "SET_POSTS":
            return {
                ...state,
                posts: action.posts
            };
        case "POST_FAIL":
            return {
                ...state,
                error: action.error,
                loading: false
            };
        case "POST_SUCCESS":
            return {
                ...state,
                error: null,
                loading: false
            };
        case "SET_INTEREST_RANDOM_POSTS":
            return {
                ...state,
                interestRandomPosts: []
            };
        case "ADD_INTEREST_RANDOM_POSTS":
            return {
                ...state,
                interestRandomPosts: [...state.interestRandomPosts, action.interestPosts]
            };
        default:
            return state;
    }
};
