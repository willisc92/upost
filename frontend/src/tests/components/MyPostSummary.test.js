import { MyPostSummary } from "../../components/MyPostSummary";
import { shallow } from "enzyme";
import React from "react";
import posts from "../fixtures/posts";

test("Should render MyPostSummary", () => {
    const post = posts[0];
    const wrapper = shallow(<MyPostSummary post={post} />);
    expect(wrapper).toMatchSnapshot();
});
