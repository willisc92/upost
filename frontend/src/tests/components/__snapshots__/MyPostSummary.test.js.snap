// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render MyPostSummary 1`] = `
<Link
  className="list-item"
  replace={false}
  to={
    Object {
      "pathname": "/myPosts/1/edit",
      "state": Object {
        "post": Object {
          "channel": 1,
          "cost": 0,
          "deleted_flag": false,
          "email": "willisc92@hotmail.com",
          "phone_number": "4034010115",
          "post_description": "Test",
          "post_event": null,
          "post_id": 1,
          "post_timestamp": "2019-06-04T23:23:53.199984-06:00",
          "post_title": "Post 1",
          "poster_name": "Willis",
          "tags": Array [
            "Interest 1",
          ],
          "user": "willisc92",
        },
      },
    }
  }
>
  <div>
    <h3
      className="list_item__title"
    >
      Post 1
    </h3>
    <p
      className="list-item__sub-title"
    >
      Test
    </p>
    <p
      className="list-item__sub-title"
    >
      Creation Date: 
      June 4th 2019
    </p>
  </div>
</Link>
`;
